testing-unit-ts/ec2tests.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright 2016-2020, Pulumi Corporation.  All rights reserved.
+
+import * as pulumi from "@pulumi/pulumi";
+import "mocha";
+
+pulumi.runtime.setMocks({
+    newResource: function(type: string, name: string, inputs: any): {id: string, state: any} {
+        switch (type) {
+            case "aws:ec2/securityGroup:SecurityGroup":
+                return {
+                    id: "sg-12345678",
+                    state: {
+                        ...inputs,
+
+                        arn: "arn:aws:ec2:us-west-2:123456789012:security-group/sg-12345678",
+                        name: inputs.name || name + "-sg",
+                    },
+                };
+            case "aws:ec2/instance:Instance":
+                return {
+                    id: "i-1234567890abcdef0",
+                    state: {
+                        ...inputs,
+
+                        arn: "arn:aws:ec2:us-west-2:123456789012:instance/i-1234567890abcdef0",
+                        instanceState: "running",
+                        primaryNetworkInterfaceId: "eni-12345678",
+                        privateDns: "ip-10-0-1-17.ec2.internal",
+                        publicDns: "ec2-203-0-113-12.compute-1.amazonaws.com",
+                        publicIp: "203.0.113.12",
+                    },
+                };
+            default:
+                return {
+                    id: inputs.name + "_id",
+                    state: {
+                        ...inputs,
+                    },
+                };
+        }
+    },
+    call: function(token: string, args: any, provider?: string) {
+        return args;
+    },
+});
+
+// It's important to import the program _after_ the mocks are defined.
```

Comment:
This is not actually valid - by spec ESM `import` and `export` evaluate before any other code, so it's not possible to actually import after running `pulumi.runtime.setMocks`. This only works because `typescript` doesn't follow spec and leaves the import the same place in the code.

[Playground Link](https://www.typescriptlang.org/play/?module=1&ssl=1&ssc=1&pln=11&pc=29#code/PTAEGEHsAcE8CcCWBzAFgF1AJgAwEYA2AWl1wBpQAFAVwBtqBbRCSeaVgQ3UUgDsA6UKACCtWqCRp0AZwkBTaXPgA3OQBN+AKE2IG7eJgBUoDrOh1GzAGbxIDUACIAAufpNgryw4DcOva0wHBkgAY1QOH21PJn54al5uBjl+RXQAWVCAa2kACgBvAF8ASm0QUABJdAByWV19dA4E0HRIUDqA5tQ5UGhbZHgOewB9Dit0JSHO7uCQ7JN4brU5K0RedS12g1BjUzbeGw5QGztHfmBVpYAPSM0QvmlMRRUlUABePYOUpVV4byA)

Babel _does_ hoist this, meaning the code wouldn't work as intended (but it would work more according to spec).

[Babel REPL](https://babeljs.io/en/repl#?code_lz=PTAEGEHsAcE8CcCWBzAFgF1AJgAwEYA2AWl1wBpQAFAVwBtqBbRCSeaVgQ3UUgDsA6UKACCtWqCRp0AZwkBTaXPgA3OQBN-AKE2IG7eJgBUoDrOh1GzAGbxIDUACIAAufpNgryw4DcOva0wHBkgAY1QOH21PJn54al5uBjl-RXQAWVCAa2kACgBvAF8ASm0QUABJdAByWV19dA4E0HRIUDqA5tQ5UGhbZHgOewB9Dit0JSHO7uCQ7JN4brU5K0RedS12g1BjUzbeGw5QGztHfmBVpYAPSM0QvmlMRRUlUABePYOUpVV4byA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=es2015%2Ctypescript%2Cenv&version=7.9.0)

If the transpiled code is kept as ESM rather than transpiled to CJS and run in a version of node that natively support ESM the tests would fail.

---

You need to either use `import("./index")` or `require("./index")` directly.

Reply:
That said, super stoked we can start writing tests for this! 👏

Reply:
@SimenB Oh, that's a great insight, thank you! Any idea of how to structure the code in a spec-compliant way and still make sure the order is preserved with the current mocks implementation?

Reply:
Yeah, as mentioned at the end

> You need to either use `import("./index")` or `require("./index")` directly.

The first one is async, so you might need something like

```ts
describe("Infrastructure", function () {
  let infra: typeof import("./index");

  before((done) => {
    import("./index")
      .then(function (resolvedInfra) {
        infra = resolvedInfra;

        done();
      })
      .catch(done);
  });

  describe("#server", function () {
    // blabla, make sure to access `infra` only inside `it`, as it hasn't been resolved outside of `it` yet
  });
});
```

(using async-await would make that example much cleaner)

Using `require` is simpler, so you should probably use that.

```ts
const infra: typeof import("./index") = require("./index");
```

Reply:
It looks like you almost have a PR ready ;)